drizzlepac/alignimages.py

```diff
@@ -82,45 +82,72 @@ def check_and_get_data(input_list,**pars):
-                imgprimaryheader = imghdu[0].header
-                try:
-                    asnid = imgprimaryheader['ASN_ID'].strip().lower()
-                    if asnid == 'NONE':
-                        asnid = None
-                except KeyError:
-                    asnid = None
-                if asnid and asnid not in retrieve_list:
+
+    # If the regression test flag is True, a regression test has been invoked where the input_list 
+    # actually contains the full filename(s) (e.g., *_flt.fits).  Note that not all regression tests use 
+    # this functionality as Artifactory or a local disk is accessed for the data source with the use of 
+    # this flag.  Some regression tests use astroquery to connect to MAST. This regression testing code 
+    # is not a scalable solution, but it is necessary for the alignment portion of the drizzle software.
+    if pars['regtest'] == True:
+        # Test folder containing the data - the root path is defined by the environment variable TEST_BIGDATA
+        data_path = pars['test_data_path']
+        from ci_watson.artifactory_helpers import get_bigdata
+        for infile in input_list:
+            full_file_path = get_bigdata(data_path[0],data_path[1],data_path[2],infile)
+            totalInputList.append(infile)
+    else:
+        archive = pars['archive']
+        clobber = pars['clobber']
+        retrieve_list = [] # list of already retrieved ASNs, only request each once
+        for input_item in input_list:
+            try:
+                filelist = aqutils.retrieve_observation(input_item,archive=archive,clobber=clobber)
+            except Exception:
+                filelist = []
+            if len(filelist) == 0:
+                # look for local copy of the file
+                input_rootname = input_item[:input_item.find('_')]
+                fitsfilenames = sorted(glob.glob("{}_fl?.fits".format(input_rootname)))
+                if len(fitsfilenames) > 0:
+                    imghdu = fits.open(fitsfilenames[0])
+                    imgprimaryheader = imghdu[0].header
                     try:
-                        filelist = aqutils.retrieve_observation(asnid,**pars)
-                        retrieve_list.append(asnid) # record asn's already retrieved
-                    except Exception:
-                        filelist = []
+                        asnid = imgprimaryheader['ASN_ID'].strip().lower()
+                        if asnid == 'NONE':
+                            asnid = None
+                    except KeyError:
+                        asnid = None
+                    if asnid and asnid not in retrieve_list:
+                        try:
+                            filelist = aqutils.retrieve_observation(asnid,**pars)
+                            retrieve_list.append(asnid) # record asn's already retrieved
+                        except Exception:
+                            filelist = []
+
+            if len(filelist) > 0:
```

Comment:
Length check not required:

```python
if filelist:
    # ...
```

Reply:
Not only that, but the previous `if` block establishes that `len(filelist)` is not 0 and therefore it _must_ be greater that 0 and therefore this `if` should be replaced with `else`.

Reply:
Done and see following issue/comment.

Reply:
If I just changed the `if` to `else`, then the totalInputList variable would not be updated with the last files acquired via retrieve_observation().  More importantly, we are reworking this routine as it needs to be robust to handle all the input configurations, both pipeline and standalone user.

Reply:
You are right. I should have analyzed this better. There are two issues here. You could have dropped `filelist` completely in favor of `totalInputList` (IMO, there is no need for `filelist` but it is really optional). The second issue _then_ is that `if len(filelist) > 0:` on line 144 is not needed at all: Python knows how to "add" an empty list to a list.

Reply:
For example,
https://github.com/spacetelescope/drizzlepac/blob/629028ed18dd52649c253236b2efb8f6924e93da/drizzlepac/alignimages.py#L119-L145
can be replaced with:
```python
totalInputList = []
for input_item in input_list:
    try:
        totalInputList.append(
            aqutils.retrieve_observation(input_item, archive=archive, clobber=clobber)
        )
    except Exception:
        input_rootname = input_item[:input_item.find('_')]
        fitsfilenames = sorted(glob.glob("{}_fl?.fits".format(input_rootname)))
        if fitsfilenames:
            with fits.open(fitsfilenames[0]) as hdulist:
                asnid = hdulist[0].get('ASN_ID', 'none').lower().strip()
                if asnid not in retrieve_list:
                    try:
                        totalInputList.append(aqutils.retrieve_observation(asnid, **pars))
                        retrieve_list.append(asnid) # record asn's already retrieved
                    except Exception:
                        pass
```

Reply:
By the way, why does this list need to be sorted at all? `fitsfilenames = sorted(glob.glob("{}_fl?.fits".format(input_rootname)))`?

Reply:
This routine is entirely being reworked.